### PR TITLE
[WIP] Introduce logging extensibility, JSON formatter extension

### LINKED ISSUE #1847: Add JsonFormatter support for the log
Hi,

it would be gread to have possibility to configure the JsonFormatter for the logs.
We can overwrite the configuration with a @BuildItem and StartupEvent but this is not nice. 

For example:
```java
 void onStart(@Observes StartupEvent ev) {
        ArrayList<Handler> handlers = new ArrayList(2);
        Formatter formatter = new JsonFormatter();
        ConsoleHandler handler = new ConsoleHandler(formatter);
        handler.setLevel(Level.INFO);
        handlers.add(handler);
        InitialConfigurator.DELAYED_HANDLER.setHandlers((Handler[]) handlers.toArray(EmbeddedConfigurator.NO_HANDLERS));
        log.info("The application is starting...");
    }
```

Thanks
Andrej

### LINKED ISSUE #2409: Support custom logging handlers
It would be great if Quarkus supported registration of custom logging handlers (some frameworks like [Sentry](https://sentry.io) use it as the integration point by registering custom JUL handlers - see an example [here](https://github.com/getsentry/examples/blob/master/java/java.util.logging/src/main/resources/logging.properties))

### PR DESCRIPTION
**Pull request by dmlloyd**
Fixes #1847, fixes #2409, enables #3985, enables #3166, enables #453, replaces #2348

### CODE REVIEW

**Review by stuartwdouglas — DISMISSED** [review]
The implementation looks good to me, but there are no docs or tests. This should have an entry in the logging guide, and ideally some kind of testing as well

File: core/runtime/src/main/java/io/quarkus/runtime/logging/LoggingSetupRecorder.java
Comment (ia3andy) [inline core/runtime/src/main/java/io/quarkus/runtime/logging/LoggingSetupRecorder.java]: Why is having this `finalErrorManager` necessary?

**Comment by ia3andy** [comment]
@dmlloyd thanks for making me co-author :)

I know it's not related to the PR but this method would factor the same code used 3 times:
```
 private static AsyncHandler createAsyncHandler(AsyncConfig asyncConfig, Level level, Handler handler) {
        final AsyncHandler asyncHandler = new AsyncHandler(asyncConfig.queueLength);
        asyncHandler.setOverflowAction(asyncConfig.overflow);
        asyncHandler.addHandler(handler);
        asyncHandler.setLevel(level);
        return asyncHandler;
    }
```

```
if (config.async.enable) {
                handlers.add(createAsyncHandler(config.async, config.level, handler));
```

**Comment by ia3andy** [comment]
@dmlloyd if you don't like my commit, feel free to edit/remove it ;-)

I just thought that a few things where hard to read like list mutated by methods..

**Comment by geoand** [comment]
@ia3andy since you have made changes, would you mind looking into creating a test and some documentation if @dmlloyd isn't already on it?

File: core/runtime/src/main/java/io/quarkus/runtime/logging/LoggingSetupRecorder.java
Comment (dmlloyd) [reply core/runtime/src/main/java/io/quarkus/runtime/logging/LoggingSetupRecorder.java]: Because it is captured in a lambda.  Or I mean... it used to be... oops :)

**Comment by dmlloyd** [comment]
> @dmlloyd if you don't like my commit, feel free to edit/remove it ;-)
> 
> I just thought that a few things where hard to read like list mutated by methods..

Yeah there are things that can be improved in the original code.  I think that might be best done as another PR or separate commits at least.

File: core/runtime/src/main/java/io/quarkus/runtime/logging/LoggingSetupRecorder.java
Comment (dmlloyd) [reply core/runtime/src/main/java/io/quarkus/runtime/logging/LoggingSetupRecorder.java]: Fixed.

File: docs/src/main/asciidoc/logging-guide.adoc
Comment (gastaldi) [inline docs/src/main/asciidoc/logging-guide.adoc]: no `<groupId>` ?

**Comment by ia3andy** [comment]
@dmlloyd I've started creating test, I have a problem with my JDK to run them. I'll continue on monday. Have a great WE cheers

File: docs/src/main/asciidoc/logging-guide.adoc
Comment (dmlloyd) [reply docs/src/main/asciidoc/logging-guide.adoc]: Fixed.

**Comment by ia3andy** [comment]
@dmlloyd I got stuck with the baby 😂I'll get back to this one during the week..

**Comment by dmlloyd** [comment]
No problem, I've been stuck with some things too.  Looks like there are some conflicts to iron out too now (I can take care of that).

**Comment by ia3andy** [comment]
I created an issue for testing:
- @dmlloyd could u rebase?
- @gsmet I think after rebase it's ready to be merged

**Comment by dmlloyd** [comment]
Again! sure, no problem :)

**Comment by loicmathieu** [comment]
I need JSON logging because of the unability to a lot of log tools to correclty handle multi-line logs ...
Is there a chance to see this merged soon ?

**Comment by ia3andy** [comment]
@loicmathieu it was supposed to be in 0.27.0, I guess we missed it in the rush.. @gsmet I'll rebase, can you mark it for the next one?

**Comment by gsmet** [comment]
Let's try to target 0.29 for that one. @ia3andy  will work on some tests next week.

**Comment by sean-scott-lr** [comment]
Since this was targeting 0.29.0 will this be put into 1.0 instead?

File: docs/src/main/asciidoc/logging.adoc
Comment (stuartwdouglas) [inline docs/src/main/asciidoc/logging.adoc]: This note just trails off.

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]
Other than the minor docs issue I think it is good to go

**Comment by gsmet** [comment]
@stuartwdouglas @ia3andy is planning to work on tests. I would prefer we have tests before merging.

**Comment by loicmathieu** [comment]
@dmlloyd  I start prototyping an extension to be able to send log events to Graylog or Logstash based on the log customization infrastructure you implement in this PR.

It works great, but it raises the question of naming log related extension, and you know, naming is one of the most complex task in programming ;)

You create an extension named `logging-json` that defines a log formatter for JSON output.

My extension is a log handler for GELF backend (Logstash or gGraylog), For the moment I named it `gelf-log-handler`.

If we develop several log extensions, better have them all in the same namespace, so I should rename mine `logging-gelf` or `logging-handler-gelf. If we choose the latest your extension should be renamed `logging-formatter-json` to be consistent I think ...

**Comment by dmlloyd** [comment]
I think `logging-gelf` makes sense.  I think it's unlikely that we'll ever need to disambiguate between handlers and formatters - if some technology covers both, I'd expect we'd want to ship a single extension that did both.

**Comment by schulzp** [comment]
Hi thanks your work so far, I'm looking forward to this feature. @dmlloyd is @stuartwdouglas' [comment](https://github.com/quarkusio/quarkus/pull/4121/files/4326b8faa059a3b1626a0969a4818a564070696d#r343900094) the only thing blocking this PR?

**Comment by dmlloyd** [comment]
Yeah as far as I know it just needs some tests.

**Comment by ia3andy** [comment]
@schulzp @dmlloyd, as commented above, I am on it. Monday was a holiday in France.

**Comment by gsmet** [comment]
For the record, this won’t be included in 1.0.0.Final so we have time to include it properly. No need to rush.

**Comment by ia3andy** [comment]
Moved to https://github.com/quarkusio/quarkus/pull/5585

**Pull request closed without merging**
